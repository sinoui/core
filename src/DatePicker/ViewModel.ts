enum ViewModel {
  dates,
  years,
  months,
  time,
}

export default ViewModel;
