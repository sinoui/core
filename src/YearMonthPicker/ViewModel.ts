enum ViewModel {
  years,
  months,
}

export default ViewModel;
