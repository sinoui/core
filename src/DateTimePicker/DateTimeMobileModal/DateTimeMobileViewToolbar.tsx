import React from 'react';
import styled from 'styled-components';
import ViewModel from '@sinoui/core/DatePicker/ViewModel';
import ClickableText from './ClickableText';
import formatTime from './formatTime';

interface Props {
  /**
   * 年份
   */
  year?: number;
  /**
   * 月份
   */
  month?: number;
  /**
   * 天
   */
  day?: number;
  /**
   * 小时
   */
  hour?: string;
  /**
   * 分钟
   */
  minute?: string;
  /**
   * 当前日历状态
   */
  viewModel: ViewModel;
  /**
   * 选择状态发生改变时的回调函数
   */
  onViewModelChange?: (viewModel: ViewModel) => void;
}

const DateTimeMobileViewToolbarWrapper = styled.div`
  height: 96px;
  padding: 16px;
  display: flex;
  justify-content: space-between;
  background-color: ${({ theme }) => theme.palette.primary.main};
  color: ${({ theme }) => theme.palette.primary.contrastText};
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;

  & > * {
    color: currentColor;
  }
`;

const LeftDateWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: space-between;
`;

const Time = styled(ClickableText)`
  align-self: flex-end;
  margin-bottom: -6px;
`;

export default function DateTimeMobileViewToolbar(props: Props) {
  const {
    year,
    viewModel,
    onViewModelChange,
    month = new Date().getMonth(),
    day = new Date().getDate(),
    hour,
    minute,
  } = props;

  const onDateClick = (type: ViewModel) => {
    if (onViewModelChange) {
      onViewModelChange(type);
    }
  };
  return (
    <DateTimeMobileViewToolbarWrapper className="sinoui-date-time-mobile-view__toolbar">
      <LeftDateWrapper>
        <ClickableText
          onClick={() => onDateClick(ViewModel.years)}
          selected={viewModel === ViewModel.years}
        >
          {year}
        </ClickableText>
        <ClickableText
          fontSize={34}
          onClick={() => onDateClick(ViewModel.dates)}
          selected={viewModel === ViewModel.dates}
        >
          {month + 1}月{day}日
        </ClickableText>
      </LeftDateWrapper>
      <Time
        fontSize={48}
        onClick={() => onDateClick(ViewModel.time)}
        selected={viewModel === ViewModel.time}
      >
        {formatTime(hour)}:{formatTime(minute)}
      </Time>
    </DateTimeMobileViewToolbarWrapper>
  );
}
