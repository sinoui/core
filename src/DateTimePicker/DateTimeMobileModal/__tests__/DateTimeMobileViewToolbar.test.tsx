import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { defaultTheme } from '@sinoui/theme';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import ViewModel from '@sinoui/core/DatePicker/ViewModel';
import DateTimeMobileViewToolbar from '../DateTimeMobileViewToolbar';

afterEach(cleanup);

describe('镜像测试', () => {
  it('渲染日期时间选择头部组件', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <DateTimeMobileViewToolbar
            viewModel={ViewModel.dates}
            year={2020}
            month={6}
            day={21}
            hour="15"
            minute="39"
          />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('验收测试', () => {
  it('渲染日期时间选择头部组件', () => {
    const { getByText } = render(
      <ThemeProvider theme={defaultTheme}>
        <DateTimeMobileViewToolbar
          viewModel={ViewModel.dates}
          year={2020}
          month={6}
          day={21}
          hour="15"
          minute="39"
        />
      </ThemeProvider>,
    );

    expect(getByText('2020')).toBeInTheDocument();
    expect(getByText('7月21日')).toBeInTheDocument();
    expect(getByText('15:39')).toBeInTheDocument();
  });

  it('分别点击年份、日期、时间，切换成对应视图', () => {
    const onViewModelChange = jest.fn();
    const { getByText } = render(
      <ThemeProvider theme={defaultTheme}>
        <DateTimeMobileViewToolbar
          viewModel={ViewModel.dates}
          year={2020}
          month={6}
          day={21}
          hour="15"
          minute="39"
          onViewModelChange={onViewModelChange}
        />
      </ThemeProvider>,
    );

    fireEvent.click(getByText('2020'));
    expect(onViewModelChange).toBeCalledWith(ViewModel.years);

    fireEvent.click(getByText('7月21日'));
    expect(onViewModelChange).toBeCalledWith(ViewModel.dates);

    fireEvent.click(getByText('15:39'));
    expect(onViewModelChange).toBeCalledWith(ViewModel.time);
  });
});
