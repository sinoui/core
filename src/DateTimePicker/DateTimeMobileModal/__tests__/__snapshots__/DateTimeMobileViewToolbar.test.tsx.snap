// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`镜像测试 渲染日期时间选择头部组件 1`] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0px;
  padding: 0px;
  margin: 0;
  border-radius: 0;
  vertical-align: middle;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  pointer-events: default;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2::-moz-focus-inner {
  border-style: none;
}

.c2:hover {
  background-color: rgba(0,0,0,0.04);
}

.c3 {
  font-size: 16px;
  opacity: 0.7;
}

.c4 {
  font-size: 34px;
  opacity: 1;
}

.c0 {
  height: 96px;
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: #3f51b5;
  color: #fff;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.c0 > * {
  color: currentColor;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 {
  font-size: 48px;
  opacity: 0.7;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  margin-bottom: -6px;
}

@media (hover:none) {
  .c2:hover {
    background-color: transparent;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <button
      className="c2 sinoui-base-button c3"
      onClick={[Function]}
      role="button"
      selected={false}
      type="button"
    >
      2020
    </button>
    <button
      className="c2 sinoui-base-button c4"
      fontSize={34}
      onClick={[Function]}
      role="button"
      selected={true}
      type="button"
    >
      7
      月
      21
      日
    </button>
  </div>
  <button
    className="c2 sinoui-base-button c5"
    fontSize={48}
    onClick={[Function]}
    role="button"
    selected={false}
    type="button"
  >
    15
    :
    39
  </button>
</div>
`;
